ipa-core/src/protocol/context/batcher.rs

```diff
@@ -1,50 +1,43 @@
-use std::{
-    collections::VecDeque,
-    sync::atomic::{AtomicUsize, Ordering},
-};
+use std::{cmp::min, collections::VecDeque, future::Future};
 
-use tokio::sync::Notify;
+use bitvec::{bitvec, prelude::BitVec};
+use tokio::sync::watch;
 
 use crate::{
+    error::Error,
     protocol::RecordId,
     sync::{Arc, Mutex},
 };
 
-pub enum Either<L, R> {
-    Left(L),
-    Right(R),
-}
-
-impl<L, R> Either<L, R> {
-    fn left(value: L) -> Self {
-        Self::Left(value)
-    }
-
-    fn right(value: R) -> Self {
-        Self::Right(value)
-    }
-}
-
 #[derive(Debug)]
-pub struct BatchState<B> {
+pub(super) struct BatchState<B> {
     pub(super) batch: B,
-    pub(super) notify: Arc<Notify>,
-    records_per_batch: usize,
-    records: AtomicUsize,
+    validation_result: watch::Sender<bool>,
+    pending_count: usize,
+    pending_records: BitVec,
 }
 
 pub(super) struct Batcher<'a, B> {
     batches: VecDeque<BatchState<B>>,
     first_batch: usize,
     records_per_batch: usize,
-    total_records: usize,
+    total_records: Option<usize>,
     batch_constructor: Box<dyn Fn(usize) -> B + Send + 'a>,
 }
 
+// Helper for `Batcher::validate_record`.
+enum Validate<B> {
```

Comment:
Should this be named something like `ValidationTime` ?

Reply:
I went with `Ready::No` and `Ready::Yes` since this is the return value from `is_ready_to_validate`.